lib/ansible/vars/reserved.py

```diff
@@ -43,35 +48,28 @@ def get_reserved_names(include_private=True):
 
         # build ordered list to loop over and dict with attributes
         for attribute in aobj.__dict__['_attributes']:
-            if 'private' in attribute:
-                private.add(attribute)
-            else:
-                public.add(attribute)
-
-    # local_action is implicit with action
-    if 'action' in public:
-        public.add('local_action')
-
-    # loop implies with_
-    # FIXME: remove after with_ is not only deprecated but removed
-    if 'loop' in private or 'loop' in public:
-        public.add('with_')
-
-    if include_private:
-        result = public.union(private)
-    else:
-        result = public
+            if not getattr(attribute, 'private', False):
+                result.add(attribute)
+            elif include_private:
+                result.add(attribute)
 
     return result
 
 
-def warn_if_reserved(myvars):
+def handle_reserved_vars(myvars):
     ''' this function warns if any variable passed conflicts with internally reserved names '''
 
-    varnames = set(myvars)
-    varnames.discard('vars')  # we add this one internally, so safe to ignore
-    for varname in varnames.intersection(_RESERVED_NAMES):
-        display.warning('Found variable using reserved name: %s' % varname)
+    if C.RESERVED_VAR_NAMES != 'ignore':
+        varnames = set(myvars)
+        for varname in varnames.intersection(_RESERVED_NAMES):
+            if varname in _RESERVE_EXCEPTIONS:
```

Comment:
create 'difference' of reserved and exceptions as constant above, slight performance increase as we already eliminate from match list